ouroboros-consensus/test-consensus/Test/Dynamic/Util.hs

```diff
@@ -221,34 +233,219 @@ leaderScheduleFromTrace (NumSlots numSlots) = LeaderSchedule .
+-- | A list of durations as number of slots. Each duration is the amount of
+-- time after a node joins the network before either the next node does or the
+-- test ends.
+--
+-- ASSUMPTIONS:
+-- * Nodes join in the order of their Id.
+-- * Node 0 joins first, at slot 0, so the first duration in the list occurs
+-- after that.
+-- * Nodes other than Node 0 join after some possibly empty duration
+-- since the previous node joined.
+-- * The last duration in the list is the run duration after the last node
+-- joins.
+-- * The number of durations is therefore the number of nodes.
+--
+-- This list is a \"weak composition\" in terms of
+-- <https://en.wikipedia.org/wiki/Composition_(combinatorics)> for the total
+-- plan duration.
+newtype ConnectionPlan = ConnectionPlan [Int]
+      -- ^ INVARIANT: all non-negative.
+      --
+      -- For example, for a test with 4 nodes and 10 slots, the durations could
+      -- be @[0,6,1,2]@ if the first two nodes join simultaneously at slot 0,
+      -- the third joins at slot 6, the fourth joins at slot 7, the full net
+      -- runs for 2 more slots, and so the test ends at slot 9 (i.e. the 10th
+      -- slot).
+    deriving (Eq)
+
+instance Show ConnectionPlan where
+  showsPrec p connPlan@(ConnectionPlan durs) = showParen (p > 10) $
+      showString "ConnectionPlan" .
+      showSpace .
+      showsPrec 0 durs .
+      showSpace .
+      showParen True
+        ( showString "ie" .
+          showSpace .
+          showsPrec 0 interp )
+    where
+      interp = (numNodes, numSlots, map unSlotNo joinSlots)
+
+      numNodes = connectionPlanNumCoreNodes connPlan
+      (joinSlots, numSlots) = interpretConnectionPlan connPlan
+
+connectionPlanNumSlots :: ConnectionPlan -> NumSlots
+connectionPlanNumSlots (ConnectionPlan durs) = NumSlots (sum durs + 1)
+
+connectionPlanNumCoreNodes :: ConnectionPlan -> NumCoreNodes
+connectionPlanNumCoreNodes (ConnectionPlan durs) = NumCoreNodes (length durs)
+
+-- | When each node joins (Core nodes Ids @0 .. n - 1@),
+-- and number of slots in the simulation.
+interpretConnectionPlan :: ConnectionPlan -> ([SlotNo], NumSlots)
+interpretConnectionPlan (ConnectionPlan durs) = (joinSlots, numSlots)
+  where
+    (durSum, joinSlots) = mapAccumL
+        (\acc dur -> (acc + dur,SlotNo (toEnum acc)))
+        0 durs
+    numSlots = NumSlots (durSum + 1)
+
+-- | The trivial \"plan\" where all nodes join immediately in slot 0.
+trivialConnectionPlan :: NumCoreNodes -> NumSlots -> ConnectionPlan
+trivialConnectionPlan (NumCoreNodes n) (NumSlots t) =
+    ConnectionPlan $ replicate (n - 1) 0 ++ [lastSlot]
+  where
+    lastSlot = t - 1
+
+-- | Inclusive minimum bound for the last duration from 'getConnectionPlan'.
+--
+-- Must be at least 1 slot so that the last node to join has an opportunity to
+-- catch up.
+--
+-- (TODO NSF someone else please review that claim.)
+minFinalDuration :: Int
+minFinalDuration = 1   {- slots -}
```

Comment:
Can you confirm this by setting it to 0 and playing with the final `threadDelay`?